Get PostgreSQL transaction tracking working

living180 opened the pull request:
There were bugs at multiple levels that prevented PostgreSQL transactions from being tracked properly.  This PR hopefully fixes all of those, plus adds a test for this functionality.

tim-schilling commented on debug_toolbar/panels/sql/tracking.py:
```diff
@@ -87,10 +100,10 @@ class NormalCursorWrapper(BaseCursorWrapper):
 
-    def __init__(self, cursor, db, logger):
+    def __init__(self, cursor, connection, logger):
         self.cursor = cursor
         # Instance of a BaseDatabaseWrapper subclass
-        self.db = db
+        self.connection = connection
```
Please keep the name `db` rather than `connection`. It should match the [cursor signature for Django](https://github.com/django/django/blob/main/django/db/backends/utils.py#L15)

living180 replied:
Sounds good.  Dropped that entire renaming commit.

tim-schilling commented on debug_toolbar/panels/sql/tracking.py:
```diff
@@ -60,10 +60,16 @@ def chunked_cursor(*args, **kwargs):
+        if connection._djdt_chunked_cursor == connection.__class__.chunked_cursor:
+            del connection.chunked_cursor
+        else:
+            connection.chunked_cursor = connection._djdt_chunked_cursor
```
How can we test that this is working as expected?

Additionally, adding summarizing the commit message would be helpful to future developers.

living180 replied:
Well, I had to implement it to get the multi-DB tests working :slightly_smiling_face:.  Not sure if it needs more testing than that...

tim-schilling replied:
That's fair, thank you!

tim-schilling commented on debug_toolbar/panels/sql/tracking.py:
```diff
@@ -213,7 +213,6 @@ def _record(self, method, sql, params):
                         "trans_id": trans_id,
                         "trans_status": final_trans_status,
                         "iso_level": iso_level,
-                        "encoding": conn.encoding,
```
@matthiask are we going to count this as a backwards incompatibility for any panel that might inherit from `SQLPanel`? It's been around since Pre v1.0, but I don't know of many things that inherit from `SQLPanel` or would use this.

living180 replied:
I don't think there is a backwards compatibility issue here since none of these attributes were ever being recorded, since prior to this PR the code was looking up the `vendor` attribute on the underlying DB connection rather than Django's `DatabaseWrapper`.  Thus none of the PostgreSQL-specific code was ever being run.

matthiask replied:
I'm fine with this change. The variable is completely unused inside django-debug-toolbar. The variable has been added in 72271115974463f30a2f9f8755e42acc69e204c3 11 years ago but I'm not sure it has ever been used anywhere.

tim-schilling replied:
That works for me. I wanted to make sure we made an active decision rather than letting it go silently. Thank you both!

tim-schilling reviewed (changes requested):
Thank you for the well written pull request. There's a lot of good information in the individual commits. I think summarizing some of that into comments would help future devs understand what things are doing and why, but I'm content without them for the most part.

living180 commented on tests/panels/test_sql.py:
```diff
@@ -506,3 +506,97 @@ def test_nested_template_information(self):
+
+    @unittest.skipUnless(
+        connection.vendor == "postgresql", "Test valid only on PostgreSQL"
+    )
```
So I changed the test to run on all databases, and for anything other than PostgreSQL, it ensures that no transaction information was recorded.

living180 commented on tests/panels/test_sql.py:
```diff
@@ -506,3 +506,97 @@ def test_nested_template_information(self):
+        query = self.panel._queries[0]
+        self.assertEqual(query[0], "default")
+        self.assertIsNotNone(query[1]["trans_id"])
+        self.assertTrue(query[1].get("starts_trans", False))
```
Updated the test to avoid the use of `.get()` in favor of explicit indexing or `in` tests.

living180 commented:
Thanks so much for the review.  I hope I have adequately addressed the points you raised.  I added comments to the parts that I felt were non-obvious from the code.  If there is anything still unclear, either from the comments that I added or that I have not yet commented, please let me know.

matthiask reviewed (approved):
It's a lot to review. I think the code looks good (better than before), tests pass and I've tested it a bit in a local environment. 

I am fine with merging this but I'll wait for a moment to see if @tim-schilling has some reservations.

tim-schilling reviewed (approved):
Thank you for the additional comments. The code makes more sense to me and is in a much better spot. This is excellent work!

living180 commented:
Thanks @matthiask and @tim-schilling for the reviews!  One question for future reference - would it be helpful if I included relevant change log entries with future PRs?

matthiask commented:
> Thanks @matthiask and @tim-schilling for the reviews! One question for future reference - would it be helpful if I included relevant change log entries with future PRs?

Sure! This would be helpful.
